File: PyScripts/Stat 8810 Examples/Testing other models.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Walmart data, but a different model (bt, binomial, poisson, hbart, probit, etc.)
1. bt: works fine but isn't very accurate at all; Update: now no mdraws
2. binomial: no mdraws
3. poisson: no mdraws
4. bart: regular (default model)
5. hbart: works fine
6. probit: no mdraws
7. modifiedprobit: no mdraws
8. merck_truncated: no mdraws
"""
import numpy as np
import matplotlib.pyplot as plt
# from sklearn import datasets, svm
# Janky importing from openbt-python repo (you'll have to change this for your own machine):
import sys
sys.path.append("/home/clark/Documents/OpenBT/openbt-python") # os.getcwd() to check
from openbt2 import OPENBT # I made changes to openbt.py & called it openbt2
sys.path.append("/home/clark/Documents/OpenBT/PyScripts/Walmart Example")
from Construct_Walmart_Data import *
from summarize_output import *

# Load in the data (8 x variables, after I edited it):
(x, y, x_pd, y_pd) = get_walmart_data()

# Settings:
# Tree prior
alpha = 0.95 # Default = 0.95
beta = 2 # Default = 2
# MCMC settings: not looking for an accurate fit yet, just testing!
N = 100 # (AKA ndpost); Default = 1000
burn = 100 # (AKA nskip); Default = 100
nadapt = 100 # default = 1000
adaptevery = 50 # Default = 100
ntreeh = 1 # Default = 1
tc = 5 # Default = 2, but we will use 5 or 6 from now on
shat = np.std(y, ddof = 1)
m=1
k=1
nu=1
nc=100
npred_arr = 40000 # Rows of the preds x grid
preds_list = []; np.random.seed(88)

# Categorical variables:
for col in [0, 1, 2, 3]:
     preds_list.append(np.random.randint(np.min(x[:, col]), np.max(x[:, col])+1, size = npred_arr))
# Continuous variables:
for col in [4, 5, 6, 7]:
     preds_list.append(np.random.uniform(np.min(x[:, col]), np.max(x[:, col])+2.2e-16, size = npred_arr))
preds_test = np.asarray(preds_list).T # This is supposedly faster than having it be a np array the whole time
# print(preds.nbytes / 1000000) # To view storage size in MB

m_test = OPENBT(model="bt", ndpost=N, nadapt = nadapt, nskip=burn, power=beta,
             base=alpha, tc=tc, numcut=nc, ntree=m, ntreeh=ntreeh, k=k,
             overallsd=shat, overallnu=nu)
fit_test = m_test.fit(x,y)
# Now, print the folder where all the config stuff is, so I can inspect:
print(m_test.fpath)
fitp_test = m.predict(preds_test)
# summarize_fitp(fitp_test)


# Next step: See how good the fit is for this "other model":     
"""
# For plotting (not accurate yet):
npreds = 100 # Default = 100
fig = plt.figure(figsize=(10,5.5))
plt.rcParams['axes.labelsize'] = 18; plt.rcParams['axes.titlesize'] = 22;
plt.rcParams['xtick.labelsize'] = 16; plt.rcParams['ytick.labelsize'] = 16;
path = 'Documents/OpenBT/PyScripts/Plots/' # Will be different for your filesystem
fname = 'Testing_Models.png'

# Plot the fitp:
ax = fig.add_subplot(111)
ax.plot(design, y, 'ro') # label=current_time can help with plot versions
ax.set_title(f'Predicted mean response +/- 2 s.d., ntree = {m.ntree}')
# ^ Technically the +/- will be 1.96 SD
ax.set_xlabel('Observed'); ax.set_ylabel('Fitted'); ax.set_ylim(-1.5, 1)
# plt.legend(loc='upper right')
# Plot the central line: Overall mean predictor
ax.plot(preds, m.mmean, 'b-', linewidth=2)

# Make the full plot with the gray lines: (mmeans and smeans are now returned by m.predict()!)
ax.plot(preds, m.mmean - 1.96 * m.smean, color='black', linewidth=0.8)
ax.plot(preds, m.mmean + 1.96 * m.smean, color='black', linewidth=0.8)
if (N < npreds):
     print('Number of posterior draws (ndpost) are less than the number of', 
                'x-values to predict. This is not recommended.')
     npreds = N
for i in range(npreds):
     ax.plot(preds, m.mdraws[i, :], color="gray", linewidth=1, alpha = 0.20)
     plt.savefig(f'{path}{fname}')
"""

# Vartivity:
fitv_test = m_test.vartivity()
# summarize_fitv(fitv)

# Sobol:
fits_test = m_test.sobol(cmdopt = 'MPI', tc = tc)
# summarize_fits(fits)

File: RScripts/Walmart Example/Testing other models.R
source("~/Documents/OpenBT/RScripts/Walmart Example/Construct_Walmart_Data.R")
source("~/Documents/OpenBT/RScripts/Walmart Example/summarize_output.R")
source("~/Documents/OpenBT/RScripts/Walmart Example/walmart_pred_plot.R")
source("~/Documents/OpenBT/openbt/Ropenbt/R/openbt2.R") # Make sure to use my local copy
# of the source code for the test of this function. this line replaces "library(Ropenbt)"
data <- get_walmart_data()
x <- data$x; y <- data$y

# Settings:
# Tree prior
alpha = 0.95 # Default = 0.95
beta = 2 # Default = 2
# MCMC settings: not looking for an accurate fit yet, just testing!
N = 100 # (AKA ndpost); Default = 1000
burn = 100 # (AKA nskip); Default = 100
nadapt = 100 # Default = 1000
adaptevery = 50 # Default = 100
ntreeh = 1 # Default = 1
tc = 5 # Default = 2, but we will use 5 or 6 from now on
shat = sd(y)
m=1
k=1
nu=1
nc=100

npred_arr = 1000; set.seed(88)
preds_test <- data.frame()
# Categorical variables:
for (v in 1:4){
  preds_test <- rbind(preds_test, sample(min(x[, v]):max(x[, v]), npred_arr, replace = TRUE))
}
# Continuous variables:
for (v in 5:8){
  preds_test <- rbind(preds_test, runif(npred_arr, min(x[, v]), max(x[, v])))
}
preds_test <- as.data.frame(t(as.matrix(preds_test))); rownames(preds_test) <- NULL
library(pryr); object_size(preds_test) # Should be only < 1 MB (not 100+!)

# rgy = c(-2, 2); fmean.out = NA
fit_test <- openbt(x, y, model="bart", ndpost=N, nadapt=nadapt, nskip=burn, power=beta,
              base=alpha, tc=tc, numcut=nc, ntree=m, ntreeh=ntreeh, k=k, 
              overallsd=shat, overallnu=nu, pbd=c(0.7,0.0))

# bt (1) needed me to specify rgy and fmean.out outside of the function, hmmm
# Also bt hangs up on the influence part of the code. These errors MAY have been
# because it doesn't let you type a string to set the first 3 models

# Diagnostics to find out why the different models aren't working in Python:
print(fit_test$folder)
# Plan: Go to the folder and view the config (which will be not-removed b/c of my source code edits)

fitp_test <- predict.openbt(fit_test, x.test=preds_test, tc=tc)

# Plot y vs yhat plots:
source("~/Documents/OpenBT/RScripts/Walmart Example/walmart_pred_plot.R")
c(ys, yhats) <- set_up_plot(fitp_test, x, y, points = 2000, var = c(1, 2, 3, 4), day_range = 30)



# Vartivity:
fitv_test = vartivity.openbt(fit_test)
# summarize_fitv(fitv)

# Sobol:
fits_test = sobol.openbt(fit_test, tc = tc)
# summarize_fits(fits)